src/python/pants/help/help_printer.py

```diff
@@ -284,7 +287,43 @@ def _print_all_tools(self) -> None:
 
     def _print_all_api_types(self) -> None:
         self._print_title("Plugin API Types")
-        print("TODO: #14227 will bring this back.")
+        api_type_descriptions: Dict[str, Tuple[str, str]] = {}
+        indent_api_summary = 0
+        for api_info in self._all_help_info.name_to_api_type_info.values():
+            name = api_info.name
+            if name.startswith("_"):
+                continue
+            if api_info.is_union:
+                name += " <union>"
```

Comment:
What's this for?

Reply:
It marks if a particular type is a union type, as that affects how that type is intended to be used, that is good to draw some attention to, imo.

Example for `PublishRequest` for instance:
<img width="786" alt="image" src="https://user-images.githubusercontent.com/72965/172234477-0d3417de-70a4-475e-8c10-e174920ccf9b.png">

And for the type specific help output the union members are included:
<img width="694" alt="image" src="https://user-images.githubusercontent.com/72965/172234718-e51ffb6f-a9b9-492d-88e7-362ef1efd3ff.png">

and for union members, the union type they are a member of is shown:
<img width="580" alt="image" src="https://user-images.githubusercontent.com/72965/172234855-2dfbd9ce-8908-4fef-a31c-39a7f9a50442.png">